Doc/c-api/init.rst

```diff
@@ -2226,6 +2226,12 @@ is resumed, and its locks reacquired.  This means the critical section API
 provides weaker guarantees than traditional locks -- they are useful because
 their behavior is similar to the :term:`GIL`.
 
+Critical sections are appropriate for protecting custom types defined by your
+C-API extensions. They should generally not be used with built-in types like
+:c:struct:`PyDictObject` or :c:struct:`PyListObject` because their public APIs
+already use critical sections internally, with the notable exception of
+:c:func:`PyDict_Next`.
```

Comment:
It's not clear to me what the "notable exception" means here. Is it saying that `PyDict_Next` *doesn't* use a critical section internally, so you should acquire the dict's critical section in client code that checks for the dict changing size to see if the iterator needs to be invalidated?

Or does it mean something else? Is there an appropriate cross-reference that could make this clearer?

Reply:
> Is there an appropriate cross-reference that could make this clearer?

I think this is referring to rhe last sentence and code example in https://docs.python.org/3.13/c-api/dict.html#c.PyDict_Next

Reply:
Yeah, `PyDict_Next ` doesn't use a critical section internally so you use `Py_BEGIN_CRITICAL_SECTION` to lock the dictionary while iterating over it. The relevant cross-reference is the [last paragraph](https://cpython-previews--124802.org.readthedocs.build/en/124802/c-api/dict.html#c.PyDict_Next:~:text=The%20function%20is%20not%20thread%2Dsafe%20in%20the%20free%2Dthreaded%20build%20without%20external%20synchronization.%20You%20can%20use%20Py_BEGIN_CRITICAL_SECTION%20to%20lock%20the%20dictionary%20while%20iterating%20over%20it%3A) of `PyDict_Next`. I don't know if we can link to that directly.

Reply:
Maybe:

> ...with the notable exception of PyDict_Next(), which requires external locking.